cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/BatchSQLEngineAdapter.java

```diff
@@ -581,11 +584,24 @@ public void waitForJobAndHandleException(SQLEngineJob<?> job) throws SQLEngineEx
   private void countRecordsIn(SQLDataset dataset,
                               @Nullable StageStatisticsCollector stageStatisticsCollector,
                               StageMetrics stageMetrics) {
+    countRecordsIn(dataset.getNumRows(), stageStatisticsCollector, stageMetrics);
+  }
+
+  /**
+   * Method to aggregate input metrics.
+   *
+   * @param numRows                  the number of input rows
+   * @param stageStatisticsCollector Stage Statistics Collector for this stage
+   * @param stageMetrics             Metrics for this stage
+   */
+  private void countRecordsIn(long numRows,
+                              @Nullable StageStatisticsCollector stageStatisticsCollector,
+                              StageMetrics stageMetrics) {
     // Count input metrics
     if (stageStatisticsCollector != null) {
-      stageStatisticsCollector.incrementInputRecordCount(dataset.getNumRows());
+      stageStatisticsCollector.incrementInputRecordCount(numRows);
     }
-    countStageMetrics(stageMetrics, Constants.Metrics.RECORDS_IN, dataset.getNumRows());
+    countStageMetrics(stageMetrics, Constants.Metrics.RECORDS_IN, numRows);
```

Comment:
At some point we should do real-time metrics by watching data push/pull and the jobs